include/swift/SIL/BasicBlockUtils.h

```diff
@@ -148,6 +148,11 @@ void findJointPostDominatingSet(
     function_ref<void(SILBasicBlock *)> foundJointPostDomSetCompletionBlocks,
     function_ref<void(SILBasicBlock *)> inputBlocksInJointPostDomSet = {});
 
+#ifndef NDEBUG
```

Comment:
I guess you have put this under `#ifndef NDEBUG` to ensure it is only used in asserts, because of its complexity. But we are shipping assert enabled compilers, e.g. for linux. Therefore we have to make sure to not introduce complexity bugs in assert statements.
Maybe it's better to not put it under NDEBUG and clearly document it's complexity.

Reply:
@eeckstein all valid points. I'm only using this as a convenient/temporary bug-finding tool. Note that we don't want to call it `dominates` because it should never ever be used when false is a valid result.

Does this comment make that more clear?

```
#ifndef NDEBUG
/// Check that \p sourceBlock dominates \p destBlock given that some path from
/// \p sourceBlock to \p destBlock exists.
///
/// Useful for *temporary* assertions when Dominance is unavailable. This is
/// worst case O(numberOfBlocksInFunction). It should only be used when \p
/// sourceBlock is expected to be "close to" \p destBlock in almost all
/// cases. Because of the potential for quadratic behavior, it should only be
/// used during feature development, never as a permanent check. If a dominance
/// check is required for correctness, then the pass should use
/// DominanceAnalysis.
bool
swift::checkDominates(SILBasicBlock *sourceBlock, SILBasicBlock *destBlock) {
  SILBasicBlock *entryBlock = sourceBlock->getParent()->getEntryBlock();
  BasicBlockWorklist worklist(destBlock);
  bool reaches = false;
  while (SILBasicBlock *block = worklist.pop()) {
    if (block == sourceBlock) {
      reaches = true;
      continue;
    }
    if (block == entryBlock) {
      return false; // does not dominate
    }
    for (auto *predBlock : block->getPredecessorBlocks()) {
      worklist.pushIfNotVisited(predBlock);
    }
  }
  return reaches;
}
#endif
```

Reply:
> Just one question: why "given that some path from \p sourceBlock to \p destBlock exists."?
> It will return false if there is no path from source to dest.

Oh thanks! I meant to remove that line from the comment